lib/action_view/template/handlers/markdown.rb

```diff
@@ -0,0 +1,29 @@
+module ActionView
+  module Template::Handlers
+    class Markdown
+      def call(_template, source)
+        @markdown ||= Redcarpet::Markdown.new(Redcarpet::Render::HTML.new(with_toc_data: true))
+        "
+        content_for :html_list_of_headings_links do
+          #{wrap_in_a_div(rendered_table_of_contents_for(source)).inspect}.html_safe
+        end
+        #{wrap_in_a_div(@markdown.render(source)).inspect}.html_safe
+        "
+      end
+
+    private
+
+      def rendered_table_of_contents_for(source)
+        @table_of_contents = Redcarpet::Markdown.new(Redcarpet::Render::HTML_TOC.new(nesting_level: 2))
+        @table_of_contents
+          .render(source)
+          .html_safe
+      end
+
+      # this makes it easier to target the generated code with CSS
+      def wrap_in_a_div(source)
+        "<div class=\"app-styled-content\">#{source}</div>"
```

Comment:
One for later probably, but: You can add a custom RedCarpet renderer to add the required classes to generated markdown, meaning you wouldn't need the app-styled-content wrapper – that's how Find and Apply are doing it.

Reply:
Yep, I guess the trade-off here is whether we want to carry extra Ruby code to generate the right classes (which will add up to quite a few LOC to cover all HTML elements), or whether we want to carry extra styles to style the generated markdown appropriately. Not sure which is better...

Reply:
I think this is probably the right tradeoff for now. We can refactor later if needed.